File: server/services/grid_service.ts
import { GridRate, PriceForecast, PricePoint } from "@shared/schema";
import { iesoClient } from "./ieso_client";

export class GridService {
  private static DEFAULT_GRID_RATE = 0.22;
  private cache: { forecast: PriceForecast; fetchedAt: number } | null = null;
  private static CACHE_MS = 10 * 60 * 1000; // 10 minutes

  async getRate(latitude: number, longitude: number): Promise<GridRate> {
    // Mock implementation fallback
    return {
      price_per_kwh: GridService.DEFAULT_GRID_RATE,
      currency: "USD"
    };
  }

  async getPriceForecast(
    latitude: number,
    longitude: number,
    horizonHours: number = 24
  ): Promise<PriceForecast> {
    // Use cached forecast when fresh
    if (this.cache && Date.now() - this.cache.fetchedAt < GridService.CACHE_MS) {
      return this.cache.forecast;
    }

    try {
      const forecast = await iesoClient.fetchDayAhead(horizonHours);
      this.cache = { forecast, fetchedAt: Date.now() };
      return forecast;
    } catch (err) {
      console.error("IESO price fetch failed, falling back to static curve:", err);
      return this.syntheticFallback(horizonHours);
    }
  }

  private syntheticFallback(horizonHours: number): PriceForecast {
    const points: PricePoint[] = [];
    for (let h = 0; h < horizonHours; h++) {
      const hourOfDay = h % 24;
      let price: number;
      let label: "peak" | "shoulder" | "off-peak";

      if (hourOfDay >= 16 && hourOfDay <= 21) {
        price = 0.40;
        label = "peak";
      } else if ((hourOfDay >= 7 && hourOfDay < 16) || (hourOfDay > 21 && hourOfDay <= 23)) {
        price = 0.22;
        label = "shoulder";
      } else {
        price = 0.12;
        label = "off-peak";
      }

      points.push({
        hour: h,
        price_per_kwh: price,
        label,
        currency: "CAD"
      });
    }
    return { points };
  }
}

export const gridService = new GridService();

File: attached_assets/extracted/develop/app/services/grid_service.py
from __future__ import annotations

import httpx
from typing import Optional, List

from app.config import settings
from app.models.schemas import GridRate, PriceForecast, PricePoint


class GridService:
    """
    Fetches local grid pricing. Falls back to DEFAULT_GRID_RATE when API is unavailable.
    """

    def __init__(self, base_url: Optional[str] = None, api_key: Optional[str] = None):
        self.base_url = base_url or settings.GRID_API_BASE_URL
        self.api_key = api_key or settings.GRID_API_KEY

    async def get_rate(
        self, latitude: float, longitude: float, use_mock: bool = False
    ) -> GridRate:
        if use_mock or not self.base_url:
            return GridRate(price_per_kwh=settings.DEFAULT_GRID_RATE)

        params = {
            "lat": latitude,
            "lon": longitude,
            "api_key": self.api_key,
        }

        async with httpx.AsyncClient(timeout=8.0) as client:
            response = await client.get(self.base_url, params=params)

        response.raise_for_status()
        data = response.json()

        # Expect a "price_per_kwh" field in dollars. Adjust to your provider shape.
        price = float(data.get("price_per_kwh") or data.get("price") or 0)
        if price <= 0:
            price = settings.DEFAULT_GRID_RATE

        return GridRate(price_per_kwh=price, currency=data.get("currency", "USD"))

    async def get_price_forecast(
        self,
        latitude: float,
        longitude: float,
        horizon_hours: int = 24,
        use_mock: bool = False,
    ) -> PriceForecast:
        if use_mock or not self.base_url:
            points: List[PricePoint] = []
            for h in range(horizon_hours):
                hour_of_day = h % 24
                if 16 <= hour_of_day <= 21:
                    price = 0.40
                    label = "peak"
                elif 7 <= hour_of_day < 16 or 21 < hour_of_day <= 23:
                    price = 0.22
                    label = "shoulder"
                else:
                    price = 0.12
                    label = "off-peak"
                points.append(
                    PricePoint(hour=h, price_per_kwh=price, label=label, currency="USD")
                )
            return PriceForecast(points=points)

        # If the real API supports forecasting, adjust below. For now, fall back to a flat rate from get_rate.
        base = await self.get_rate(latitude=latitude, longitude=longitude, use_mock=False)
        points = [
            PricePoint(hour=h, price_per_kwh=base.price_per_kwh, label="shoulder", currency=base.currency)
            for h in range(horizon_hours)
        ]
        return PriceForecast(points=points)
